loopy/tools.py

```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+    """
+    from mako.template import Template
+    from loopy.kernel.instruction import MultiAssignmentBase, BarrierInstruction
+
+    options = {}  # options: mapping from insn_id to str of options
+
+    for insn in kernel.instructions:
+        option = f"id={insn.id}, "
+        if insn.depends_on:
+            option += ("dep="+":".join(insn.depends_on)+", ")
+        if insn.tags:
+            option += ("tags="+":".join(insn.tags)+", ")
+
+        if isinstance(insn, MultiAssignmentBase):
+            if insn.atomicity:
+                option += "atomic, "
+        elif isinstance(insn, BarrierInstruction):
+            option += (f"mem_kind={insn.mem_kind}, ")
+        else:
+            pass
+
+        options[insn.id] = option[:-2]  # get rid of the trailing ", "
+
+    python_code = r"""
+    <%! import loopy as lp %>
+    import loopy as lp
+    import numpy as np
+    from pymbolic.primitives import *
+
+    <%! tv_scope = {0: 'lp.AddressSpace.PRIVATE', 1: 'lp.AddressSpace.LOCAL',
+    2: 'lp.AddressSpace.GLOBAL', lp.auto: 'lp.auto' } %>
+    ${var_name} = lp.make_kernel(
+        [
+        % for dom in kernel.domains:
+        "${str(dom)}",
+        % endfor
+        ],
+        '''
+        % for id, opts in options.items():
+        <% insn = kernel.id_to_insn[id] %>
+        % if isinstance(insn, lp.Assignment):
+        ${insn.assignee} = ${insn.expression} {${opts}}
+        % elif isinstance(insn, lp.BarrierInstruction):
+        ... ${insn.synchronization_kind[0]}barrier{${opts}}
+        % elif isinstance(insn, lp.NoOpInstruction):
+        ... nop {${opts}}
+        % else:
+        <% raise NotImplementedError("Not implemented for ${type(insn)}.")%>
+        % endif
+        %endfor
+        ''', [
+            % for arg in kernel.args:
+            % if isinstance(arg, lp.ValueArg):
+            lp.ValueArg(
+                name="${arg.name}",
+                dtype=${('np.'+arg.dtype.numpy_dtype.name
+                            if arg.dtype else 'lp.auto')}),
+            % else:
+            lp.GlobalArg(
+                name="${arg.name}", dtype=${('np.'+arg.dtype.numpy_dtype.name
+                                                if arg.dtype else 'lp.auto')},
+                shape=${arg.shape}, for_atomic=${arg.for_atomic}),
+            % endif
+            % endfor
+            % for tv in kernel.temporary_variables.values():
+            lp.TemporaryVariable(
+                name="${tv.name}",
+                dtype=${'np.'+tv.dtype.numpy_dtype.name if arg.dtype else 'lp.auto'},
+                shape=${tv.shape}, for_atomic=${tv.for_atomic},
+                address_space=${tv_scope[tv.address_space]},
+                read_only=${tv.read_only},
+                % if tv.initializer is not None:
+                initializer=${"np."+repr(tv.initializer)},
+                % endif
+                ),
+            % endfor
+            ],
+            lang_version=${lp.MOST_RECENT_LANGUAGE_VERSION},
```

Comment:
Why not reproduce the original language version?

Reply:
Loopy kernels do not store language versions. The only way to reproduce lang version would be by copying the entire `LoopKernel.options`, I think that wouldn't be quite good looking.